Migrate to NextJS Routes

SudharakaP opened the pull request:
Resolves https://github.com/opencollective/opencollective/issues/3791

# Description

@Betree : I am opening this PR to incrementally migrate to NextJS routes. Here I have migrated the `updates.js` page to NextJS routes. Let me know if the approach seems okay and I will slowly migrate the other pages as well one by one. Ultimately I think we'll be able to delete `pages.js` and also probably get rid of `Link.js`. 😄 

I think the other PR I opened (https://github.com/opencollective/opencollective-frontend/pull/5783) is too messy as it tries to migrate all at once. And I prefer this approach. I'll close that one later.

SudharakaP commented:
@Betree : I'll keep checking in the changes to each page in single commits so that the reviewing is easier perhaps. Let me know if you prefer a different method. 😄

SudharakaP commented:
@Betree : This draft is completed and ready for testing and review. I've tested the application as a whole and seems to work okay. Although, since this is a large change there might be few regressions somewhere which I might have missed. We probably need some more testing to be done. 🤔 

1) I've just pushed as individual commits to make the review process easier perhaps; but I don't know if it makes a difference or not. If you have any specific questions or queries feel free to ping me. 😄 

2) The approach I used here is to migrate all routes we had in the `pages.js` to `rewrites.js` and change the `Router` and `Link` objects to use `next/link` and `Router` from NextJS. Now I haven't tried to use the [directory structure of NextJS](https://nextjs.org/docs/basic-features/pages) to arrange the routes since I thought that given that some of our routes are complicated it might get messy. Also having some routes follow the directory based routing structure and some using the `rewrites.js` file didn't somehow sit with me right. Anyways let me know if you have different ideas. 😄 

3) I did remove some logic which I thought didn't make sense; specifically in `Link.js`; https://github.com/opencollective/opencollective-frontend/pull/5855/files#diff-62b3e26d6c32dd4cbe13c9c2b2edb876f7d36542273afb08dea55ec90851a5cbL50-L57. I didn't see any use of this; and the e2e tests run fine without it although it says `cypress is using an iframe for e2e testing`. Maybe legacy issue that doesn't apply anymore? 

4) There were few scenarios where I have to delete empty and null values from a json object and I used the following logic; https://github.com/opencollective/opencollective-frontend/pull/5855/files#diff-227d8b6a0b0c9a263a7e0411bae4262e49aaffa8a43cd2a8c6bccf03838ad85bR150-R153. Don't know if there's a simple or better way of doing it, I'll think more; but if you have anything off the top of your head please let me know. 😄 

Let me know your thoughts. Thanks and hope this helps. 😄

Betree reviewed (commented):
I want to do some more testing, but so far I'm quite impressed. Everything I tested so far on Zeit worked fine, even the routes with regexes (eg. https://opencollective-frontend-bj8twodx4.vercel.app/babel/order/355/500/year).

Thanks for starting that @SudharakaP!

Betree commented on components/ApplyToHostModal.js:
```diff
@@ -447,10 +447,11 @@ ApplyToHostModal.propTypes = {
 };
 
-export default ApplyToHostModal;
+export default withRouter(ApplyToHostModal);
```
:ice_cream: Not a requested change, just FYI we can also use `useRouter` for functional components

SudharakaP replied:
Yes, I was contemplating on if I should make that distinction between functional and non-functional components. But I opted for `withRouter` for simplicity so I have a bit of less work and we have the same pattern everywhere. But if you do see a usefulness in using `useRouter` specifically for functional components let me know and I'd be happy to make that change. 😄

Betree commented on components/CollectiveCard.js:
```diff
@@ -181,20 +181,15 @@ class CollectiveCard extends React.Component {
+      route = `${collective.parentCollective?.slug || 'collective'}/events/${collective.slug}`;
     } else {
-      route = 'collective';
-      params = { slug: collective.slug };
+      route = collective.slug;
```
It's important to start the URLs with a `/`. See https://opencollective-frontend-bj8twodx4.vercel.app/babel/redeem, for example, the link is rendered as: https://opencollective-frontend-bj8twodx4.vercel.app/babel/babel

SudharakaP replied:
@Betree : Are you sure? Because when I go to https://opencollective-frontend-bj8twodx4.vercel.app/babel/redeem and click on the link I get redirected to https://opencollective-frontend-bj8twodx4.vercel.app/babel.

![Peek 2021-02-15 06-48](https://user-images.githubusercontent.com/12435965/107960726-d0c0c980-6f59-11eb-9d93-096502728839.gif)

Betree replied:
The redirect indeed works, but there's definitely something going on with the `href`:

![image](https://user-images.githubusercontent.com/1556356/107978484-224e7000-6fbd-11eb-9d45-23d1ed85dd08.png)

SudharakaP replied:
@Betree : You are absolutely right indeed. All the `next/link` examples points to the fact the route should be prefixed with the `/`. So I've went ahead and done this change. Seems now it's working correctly. 😄

SudharakaP commented on components/Link.js:
```diff
@@ -47,17 +50,9 @@ class Link extends React.Component {
           {children}
         </Scrollchor>
       );
-    } else if (this.isIframe) {
-      const routeFromRouter = router.findByName(route);
-      const path = routeFromRouter ? routeFromRouter.getAs(params) : `https://opencollective.com${route}`;
-      return (
-        <a href={path} title={title} target="_top" className={className} {...otherProps}>
-          {children}
-        </a>
-      );
```
@Betree : Sorry, but I am not sure what you mean by widgets here? So these widgets are not part of our tests? 🤔

Betree replied:
The ones that projects integrate into their README or on their websites. You can preview some of them on https://opencollective.com/widgets. Unfortunately, these are based on old code and have very poor test coverage. I think the main one is `banner.html`

Betree replied:
You can see one here: https://kivy.org/#home. If you try to manually change the iframe URL to https://opencollective-frontend-bj8twodx4.vercel.app/kivy/banner.html?limit=10&id=banner-iframe-6108&style={} you'll see that it indeed creates a regression: the page should open in a new tab. So I guess we should keep this for now.

![Peek 2021-02-15 18-44](https://user-images.githubusercontent.com/1556356/107978891-d8b25500-6fbd-11eb-9e3f-45275bf85a71.gif)

SudharakaP replied:
@Betree : I tested that. But even without the url change it doesn't seem to open in a new tab anyways. So I guess that is not related to this change? 🤔 

![pre-test](https://user-images.githubusercontent.com/12435965/107985216-80a82e00-6f7e-11eb-84f6-a00c64b8f6ae.gif)

Betree replied:
You're actually right, I'll open a separate issue about that. Since this code is not playing its role, we can indeed remove it.

Betree replied:
Issue created here: https://github.com/opencollective/opencollective/issues/3984

If you see a quick fix, feel free to implement it in this PR. Otherwise, we can keep that for later

SudharakaP replied:
@Betree : Took a quick look and I couldn't figure something obvious so we'll leave it for the other issue. 🤔

SudharakaP commented on pages/expenses.js:
```diff
@@ -141,20 +142,27 @@ class ExpensePage extends React.Component {
+    Object.keys(queryParameters).forEach(
+      k => !queryParameters[k] && queryParameters[k] !== undefined && delete queryParameters[k],
+    );
+    return queryParameters;
```
@Betree : Thanks a lot for the suggestion. Just checked in changes to omit `null`, `undefined` and `empty` values using `omitBy` method. Hope that looks good. Let me know. 😄

SudharakaP commented:
> I want to do some more testing, but so far I'm quite impressed. Everything I tested so far on Zeit worked fine, even the routes with regexes (eg. https://opencollective-frontend-bj8twodx4.vercel.app/babel/order/355/500/year).
> 
> Thanks for starting that @SudharakaP!

@Betree : You are most welcome 😸

Betree commented on components/accept-financial-contributions/StripeOrBankAccountPicker.js:
```diff
@@ -177,14 +176,7 @@ class StripeOrBankAccountPicker extends React.Component {
-                      method: 'bank',
-                    }}
-                  >
+                  <Link href={`${router.query.slug}/accept-financial-contributions/${router.query.path}/bank`}>
```
```suggestion
                  <Link href={`/${router.query.slug}/accept-financial-contributions/${router.query.path}/bank`}>
```

Betree commented on components/accept-financial-contributions/StripeOrBankAccountPicker.js:
```diff
@@ -220,11 +212,9 @@ class StripeOrBankAccountPicker extends React.Component {
-                state: 'success',
-              });
+              await this.props.router.push(
+                `${router.query.slug}/accept-financial-contributions/${router.query.path}/success`,
```
```suggestion
                `/${router.query.slug}/accept-financial-contributions/${router.query.path}/success`,
```

Betree commented on components/accept-financial-contributions/SuccessPage.js:
```diff
@@ -139,11 +139,7 @@ class SuccessPage extends React.Component {
-                              slug: collective.host.slug,
-                              section: 'fiscal-hosting',
-                            }}
+                            href={`${collective.host.slug}/edit/fiscal-hosting`}
```
```suggestion
                            href={`/${collective.host.slug}/edit/fiscal-hosting`}
```

Betree commented on components/collective-page/hero/HeroEventDetails.js:
```diff
@@ -135,7 +135,7 @@ class HeroEventDetails extends React.Component {
           <HeroNote>
             <MapPin size={16} />
-            <Link route={locationRoute}>
+            <Link href={`/${locationRoute}`}>
```
No need for the `#` here, locationRoute is a hash: https://github.com/opencollective/opencollective-frontend/blob/55b9d3f688e80e01b944b5df6f0b9fc950770490/components/collective-page/hero/HeroEventDetails.js#L89

```suggestion
            <Link href={`${locationRoute}`}>
```

Betree commented on components/collective-page/sections/Events.js:
```diff
@@ -109,7 +109,7 @@ class SectionEvents extends React.PureComponent {
         {Boolean(events?.length) && (
           <ContainerSectionContent>
-            <Link route="contribute" params={{ collectiveSlug: collective.slug, verb: 'contribute' }}>
+            <Link href={`${collective.slug}/contribute`}>
```
```suggestion
            <Link href={`/${collective.slug}/contribute`}>
```

Betree commented on components/collective-page/sections/Updates.js:
```diff
@@ -127,7 +127,7 @@ class SectionUpdates extends React.PureComponent {
           </P>
           {isAdmin && (
-            <Link route="createUpdate" params={{ collectiveSlug: collective.slug }}>
+            <Link href={`${collective.slug}/updates/new`}>
```
```suggestion
            <Link href={`/${collective.slug}/updates/new`}>
```

Betree commented on components/collective-page/sections/Updates.js:
```diff
@@ -164,7 +164,7 @@ class SectionUpdates extends React.PureComponent {
                     </Box>
                     <Flex flexDirection="column" justifyContent="space-between">
-                      <Link route="update" params={{ collectiveSlug: collective.slug, updateSlug: update.slug }}>
+                      <Link href={`${collective.slug}/updates/${update.slug}`}>
```
```suggestion
                      <Link href={`/${collective.slug}/updates/${update.slug}`}>
```

Betree commented on components/collective-page/sections/Updates.js:
```diff
@@ -242,7 +237,7 @@ class SectionUpdates extends React.PureComponent {
             </StyledCard>
             {updates.length > 0 && (
-              <Link route="updates" params={{ collectiveSlug: collective.slug }}>
+              <Link href={`${collective.slug}/updates`}>
```
```suggestion
              <Link href={`/${collective.slug}/updates`}>
```

Betree reviewed (commented):
There are a bunch of `href` still not including a leading `/`. The following regex helps to find them (tried with VSCode):

```
# For links
href=\{[`'][^/#(HTTP))]

# For pushRoute
router.push\([`'][^/#(HTTP))]
```

SudharakaP commented on components/UserProvider.js:
```diff
@@ -2,11 +2,11 @@ import React from 'react';
 import PropTypes from 'prop-types';
 import { withApollo } from '@apollo/client/react/hoc';
 import { isEqual } from 'lodash';
+import Router from 'next/router';
```
@Betree : Yes this works. All one must do to verify is to put a console log statement above the `if` statement and reload the `signin page`. Here's a screencast. 😄 

![signin](https://user-images.githubusercontent.com/12435965/108083231-b9501200-7027-11eb-87a7-362d4114e991.gif)

SudharakaP commented on components/Pagination.js:
```diff
@@ -79,9 +78,8 @@ const Pagination = ({ route, limit, offset, total, scrollToTopOnChange, isDisabl
       {currentPage < totalPages && (
         <Link
-          route={route || router.route.slice(1)}
+          href={{ pathname: route || router.route.slice(1), query: { ...router.query, offset: offset + limit } }}
```
@Betree : For this one, actually I saw this initially but missed writing it in my initial post. Now the problem here is that the query object returned by the router has the full query details including the slug. 

> There are a bunch of `href` still not including a leading `/`. The following regex helps to find them (tried with VSCode):
> 
> ```
> # For links
> href=\{[`'][^/#(HTTP))]
> 
> # For pushRoute
> router.push\([`'][^/#(HTTP))]
> ```

@Betree : Sorry about this, it seems I missed some due to not using the full search in Intellij but rather a short-cut method. But now I've corrected all of them. 😄

SudharakaP replied:
@Betree : I saw this when I was testing initially but somehow missed writing it down. I was thinking whether this should be fixed or not as I don't see it as a problem? 🤔 Now in the NextJS router the query includes the slug whereas previously it didn't and was part of the pathname. If we are to fix it since we are using the router object here, I was thinking it would perhaps be to change the passed route name to include the slug and not using the `router.query` within the `query` parameter. Anyways let me know your thoughts.

Betree replied:
We do want this route to be properly formatted. We can eventually solve it in a separate issue, but it would be great if we can do it here. It's probably the same solution that we need for https://github.com/opencollective/opencollective-frontend/pull/5795.

My understanding of the problem is not good enough to give recommendations, but I can look at it tomorrow.

SudharakaP replied:
@Betree : Not to worry, I was under the impression that's a valid route but I think we are opting for routes without query parameter correct? Anyways, I've pushed my change (https://github.com/opencollective/opencollective-frontend/pull/5855/commits/889af8f694ac7ceb47cab7de4ec83d9e4b4c00f6) for that and that should take care of it. Let me know if you see any issues. 😄

SudharakaP commented:
> There are a bunch of `href` still not including a leading `/`. The following regex helps to find them (tried with VSCode):
> 
> ```
> # For links
> href=\{[`'][^/#(HTTP))]
> 
> # For pushRoute
> router.push\([`'][^/#(HTTP))]
> ```

@Betree : Sorry about this, it seems I missed some due to not using the full search in Intellij but rather a short-cut method. But now I've corrected all of them. 😄

Betree commented on components/contribute-cards/ContributeTier.js:
```diff
@@ -157,13 +149,7 @@ const ContributeTier = ({ intl, collective, tier, ...props }) => {
-                    tierSlug: tier.slug,
-                    tierId: tier.id,
-                  }}
+                  href={{ pathname: `/${collective.slug}/contribute/${tier.slug}`, query: { tierId: tier.id } }}
```
`tierId` should be part of the URL, this one doesn't work: https://opencollective-frontend-nw4onuvfg.vercel.app/captainfact_io/contribute/tshirt?tierId=10355

SudharakaP commented:
@Betree : One other thing to reiterate is that we need to find a solution to [point 5) I've highlighted above](https://github.com/opencollective/opencollective-frontend/pull/5855#issuecomment-778975026); without that the dev experience will not be the best as the warning is frequently occurring in the dev console. I will be working on that next. 🤔

SudharakaP commented:
@Betree : I've changed some things in order to get rid of warnings related to links in 3995a44. One thing I note is that going forward we should avoid using any child `<a>` tags within `<Link>` tags; directly or indirectly such as by using child components such as `<StyledLink>`. Since we already have an inner `<a>` tag in [Link.js](https://github.com/opencollective/opencollective-frontend/blob/5795472f6a5540a289a325ae78582f78da9ca567/components/Link.js#L61-L69) this will create warnings. I cleaned up most of it and ran some checks in most of the collective, profile pages etc but if you see any I've missed feel free to let me know. 🤔

Betree commented:
> @Betree : I've changed some things in order to get rid of warnings related to links in [3995a44](https://github.com/opencollective/opencollective-frontend/commit/3995a4446b2180cd32ad37ff102f56049b868730). One thing I note is that going forward we should avoid using any child `<a>` tags within `<Link>` tags; directly or indirectly such as by using child components such as `<StyledLink>`. Since we already have an inner `<a>` tag in [Link.js](https://github.com/opencollective/opencollective-frontend/blob/5795472f6a5540a289a325ae78582f78da9ca567/components/Link.js#L61-L69) this will create warnings. I cleaned up most of it and ran some checks in most of the collective, profile pages etc but if you see any I've missed feel free to let me know. thinking

You did the right thing. The rule you mention was actually already true for all links rendered with `components/Link`, it's only for the ones we rendered with the link from `../server/pages` that it was not applied.

Betree reviewed (commented):
We need to rebase before I can deploy to staging (we merged a fix for a build issue on `master`)

SudharakaP commented:
> We need to rebase before I can deploy to staging (we merged a fix for a build issue on `master`)

@Betree : I've applied changes from master. Now it should be good to go. 😄

SudharakaP commented on components/Footer.js:
```diff
@@ -365,8 +365,8 @@ const Footer = () => {
                 {Object.keys(navigation[key]).map(item => (
                   <ListItem key={item} textAlign={['center', 'left']} mb={2}>
                     {navigation[key][item][0] === '/' ? (
-                      <Link route={navigation[key][item]} passHref>
```
@Betree : Great point. I'll work on this today. 😄

SudharakaP replied:
@Betree : Done. 👍🏽

Betree commented on pages/external-redirect.js:
```diff
@@ -84,7 +84,7 @@ const ExternalRedirectPage = () => {
                 </StyledButton>
               </a>
-              <Link route={fallback} onClick={() => setPendingAction('CANCEL')}>
+              <Link href={`/${fallback}`} onClick={() => setPendingAction('CANCEL')}>
```
The one in https://github.com/opencollective/opencollective-frontend/blob/915be5e17ab6df878375e25c8c56248a4de4275a/components/contribution-flow/index.js#L232 already has a `/`, and the ones we generate from https://github.com/opencollective/opencollective-api/blob/a759154e1efcb5ef1c196f772613e7c0ba8bea4d/server/lib/sanitize-html.ts#L187 have an `https://` prefix.
```suggestion
              <Link href={fallback} onClick={() => setPendingAction('CANCEL')}>
```

Betree reviewed (commented):
I fixed a few issues. Almost everything seemed to work fine, but we're still having issues with pagination and search. Here's my full test report (after applying https://github.com/opencollective/opencollective-frontend/pull/5855/commits/6d0e72e9f461e93c5fbf3f0cf092828ebbf4d628):

- [x] Sign in / Sign up
    - [x] From dedicated page
    - [x] From contribution flow
- [x] Contribution flow redirect
- [x] Create collective
- [x] Onboarding flow
- [x] Accept contributions flow
- [x] Anchor scroll (navbar from both collective page & non-collective page)
- [x] Back button (eg. on error page, seatch for this.props.router.back)
- [x] Gift cards (esp. redeem)
- [x] Conversations
- [ ] Search (including on transactions, dashboard, etc)
    - [x] Main search
    - [x] Transactions
    - [ ] Expenses => breaks route (same with filters)
    - [ ] Host dashboard > Expenses => Error (same with filters)
    - [ ] Host dashboard > Contributions => Error (same with filters)
    - [ ] Host dashboard > Applications => Error (same with filters)
    - [ ] Host dashboard > Hosted collectives => Error (same with filters)
- [ ] Pagination
    - [ ] Breaks page on search https://staging.opencollective.com/search
    - [ ] Breaks page on all dashboard views
    - [ ] Loosing filters on /transactions
    - [ ] Loosing filters on /expenses
    - [ ] Loosing filters on /discover
- [x] WarnIfUnsavedChanges
- [x] Links to sections of the settings page (eg. "Edit fiscal hosting")
- [x] Contribute button on funds (navbar)
- [x] Contribute cards CTAs & links (tiers, tickets, events)
- [x] Static pages

SudharakaP commented on pages/host.dashboard.js:
```diff
@@ -24,9 +23,11 @@ import Loading from '../components/Loading';
 
-const MenuLink = styled(props => <Link {...omit(props, ['isActive'])} />)`
+// eslint-disable-next-line react/display-name
+const LinkContainer = styled(React.forwardRef((props, ref) => <StyledLink {...props} ref={ref} />))`
```
@Betree : I don't know why I was doing that tbh, in fact using a `StyledLink` within a `Link` tag isn't the best as that will have nested <a> tags. So I refactored this part to use a `Container` component. Also just realized the route in `host.dashboard` had a small problem and that was fixed as well. I think you highlighted that in your report.

SudharakaP commented:
@Betree : I believe I've solved all the issues you've outlined. Let me know if you find anything else. 

Now one thing that occurred to me is that given the nature of NextJS page routing; a user can go to `<base-url>/expences` and this will create an error (i.e: https://staging.opencollective.com/expenses). Notice that page level routing getting precedence over rewrites. This is indeed a valid error as we are not passing a slug here. I feel this can have unintended consequences, such as a collective name clashing with a page name (i.e: user creates a collective like, `Expenses`). Therefore I feel we need a way to make sure our `rewrites` urls doesn't clash with page names. My suggestion is to move every single custom page to a inner folder (we name it say, custom-pages within the pages folder and alter all the rewrite urls respectively.  What do you think? 🤔

SudharakaP commented:
>My suggestion is to move every single custom page to a inner folder (we name it say, custom-pages within the pages folder and alter all the rewrite urls respectively.

@Betree : Thinking on this further I don't think that's the best solution because it would add a prefix to all the routes and that I think would create problems. I guess we need to handle that in a different way. 🤔 🤔 🤔

Betree commented:
> @Betree : I believe I've solved all the issues you've outlined. Let me know if you find anything else. smile
> 
> Now one thing that occurred to me is that given the nature of NextJS page routing; a user can go to `<base-url>/expences` and this will create an error (i.e: [staging.opencollective.com/expenses](https://staging.opencollective.com/expenses)). Notice that page level routing getting precedence over rewrites. This is indeed a valid error as we are not passing a slug here. I feel this can have unintended consequences, such as a collective name clashing with a page name (i.e: user creates a collective like, `Expenses`). Therefore I feel we need a way to make sure our `rewrites` urls doesn't clash with page names. My suggestion is to move every single custom page to a inner folder (we name it say, custom-pages within the pages folder and alter all the rewrite urls respectively. What do you think? thinking

Having a collective name clashing with a page name is normally prevented in https://github.com/opencollective/opencollective-api/blob/fda9f0c8786ed996e79e77873377487b51b8a79c/server/lib/collectivelib.ts#L211.

However, it seems that we want to [disable file-system routing](https://nextjs.org/docs/advanced-features/custom-server#disabling-file-system-routing) with `useFileSystemPublicRoutes: false`, at least for now.

SudharakaP commented:
> Having a collective name clashing with a page name is normally prevented in https://github.com/opencollective/opencollective-api/blob/fda9f0c8786ed996e79e77873377487b51b8a79c/server/lib/collectivelib.ts#L211.
> 
> However, it seems that we want to [disable file-system routing](https://nextjs.org/docs/advanced-features/custom-server#disabling-file-system-routing) with `useFileSystemPublicRoutes: false`, at least for now.

@Betree : Ah, I completely missed this config. Thanks for letting me know. 😸

SudharakaP commented:
@Betree : So it seems that, 

1) `rewrites.js` had some route conflicts. So I had to move the `collective-page` all the way to the bottom so that the page level routes (`/discover`, `/search` etc) gets precedence. 

2) Also it seems that `index.js` is handled in some sort of special way and therefore I renamed it to `home.js`; otherwise it still goes to the `./collective-page`. 

These seem to solve all issues. 😸

Betree commented:
> rewrites.js had some route conflicts. So I had to move the collective-page all the way to the bottom so that the page level routes (/discover, /search etc) gets precedence.

Good catch! That's why I missed it locally, once everything was done I just thought "let's move all these static routes at the bottom to get something cleaner", but haven't re-tested after that. Collective page route should always be the last one.

I'll push to staging now to do some final testing, this is looking good!

Betree commented:
Rebasing on master

SudharakaP commented:
> > rewrites.js had some route conflicts. So I had to move the collective-page all the way to the bottom so that the page level routes (/discover, /search etc) gets precedence.
> 
> Good catch! That's why I missed it locally, once everything was done I just thought "let's move all these static routes at the bottom to get something cleaner", but haven't re-tested after that. Collective page route should always be the last one.
> 
> I'll push to staging now to do some final testing, this is looking good!

@Betree : Wonderful. Let me know if you need any help on this further. Hope everything is looking good now. 😄

SudharakaP commented:
@Betree : FYI, I've noticed that we are missing `/` on two routes and added them at 992bda7. Hope everything is good now. 🤔
